### PR TITLE
Enforce proper return types for methods get_order_from_event_body

### LINKED ISSUE #4293: Enforce proper return types for methods/functions get_order_from_event_body_intent_id
### Description

Based on this convo https://github.com/Automattic/woocommerce-payments/pull/4287#pullrequestreview-984284936

It's more than one time we see fatal errors due to mixed return types.

In this specific case #4287, `get_order_from_event_body_intent_id` returns a few types:

https://github.com/Automattic/woocommerce-payments/blob/caf5ca2298ea77ad4e2b8f8f03c3f037f0a871c6/includes/class-wc-payments-webhook-processing-service.php#L551-L553

But when a new code is added - [ref](https://github.com/Automattic/woocommerce-payments/commit/eb591d24146989485ce9ec65151e3f2e7bb4df95), it misses checking the proper type before using it.

### Acceptance criteria

* Update/refactor `get_order_from_event_body_intent_id` to return less types. Ideally, it would be `null | WC_Order`, and ensure their consumers `process_webhook_payment_intent_failed` and `process_webhook_payment_intent_succeeded` properly check their types before using.
* (Maybe) Check other functions to enforce the same. It's worth checking if PHPCS or Psalm can enforce or prevent this issue. 

### Dev notes

IIRC, Psalm is pretty smart in detecting types and it can prevent this type of issues. Quickly found out this post https://psalm.dev/articles/php-or-type-safety-pick-any-two

### Additional context

This may fit into the RFC improving the plugin code quality paJDYF-3eh-p2

### PR DESCRIPTION
**Pull request by mgascam**
Fixes #4293 

This pull request includes changes to the `WC_Payments_Webhook_Processing_Service` class and its associated tests. The changes focus on improving error handling and ensuring proper return types in the webhook processing service.

#### Changes proposed in this Pull Request
* Narrow dockblock return types to `null` or `WC_Order` of `get_order_from_event_body`
* Changes to fulfill the new return types
* Add unit test

#### Testing instructions

The `get_order_from_event_body` is used in the `succeded` and `failed` payment intent webhook management. Make sure you are listening to Stripe webhook events in your local server

**Payment intent succeeds**
1. In your local store complete a successful checkout with `4242424242424242` test card
2. Wait until the `payment_intent.suceeded` event is processed
3. Check that the order is marked as processing in WP Admin

**Payment intent fails**
1. In your local store complete a failed checkout with `4100000000000019` test card
2. Wait until the `payment_intent.failed` event is processed
3. Check that the order is marked as failed in WP Admin
-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9761` or branch name `dev/4293-enforce-proper-return-types-for-methodsfunctions-get_order_from_event_body_intent_id` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=dev/4293-enforce-proper-return-types-for-methodsfunctions-get_order_from_event_body_intent_id) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: c95bc469447d4d42ee0a3fa4b7b2104183f3d960
- Build time: 2024-11-26 12:53:36 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/class-wc-payments-webhook-processing-service.php [hunk 1]
```diff
@@ -740,11 +741,11 @@ private function get_order_from_event_body( $event_body ) {
 				$order = $this->wcpay_db->order_from_order_id( $order_id );
 			} elseif ( ! empty( $event_object['invoice'] ) ) {
 				// If the payment intent contains an invoice it is a WCPay Subscription-related intent and will be handled by the `invoice.paid` event.
-				return false;
+				return null;
 			}
 		}
 
-		if ( ! $order ) {
+		if ( ! $order instanceof \WC_Order ) {
```
Comment (zmaglica) [inline includes/class-wc-payments-webhook-processing-service.php:748]: In 99% of cases, this works fine. However, if you take a closer look at this line: `this->wcpay_db->order_from_order_id`, you'll see that `order_from_order_id` can return a `WC_Order_Refund` (according to the docblock). Since `\WC_Order` is not an instance of `WC_Order_Refund`, this would make the following statement false and potentially cause issues. Instead, you could use the abstract class `WC_Abstract_Order` for the comparison. Both `WC_Order` and `WC_Order_Refund` extend this class, which would make the statement true.

File: tests/unit/test-class-wc-payments-webhook-processing-service.php [hunk 2]
```diff
@@ -1733,4 +1733,31 @@ public function provider_mode_mismatch_detection() {
 			'No mode proceeds'                => [ null, $this->once() ],
 		];
 	}
+
+	public function test_process_throws_exception_when_order_not_found_for_intent_id() {
```
Comment (zmaglica) [inline tests/unit/test-class-wc-payments-webhook-processing-service.php]: Could you also replicate a test and add a case where `WC_Order_Refund` is returned so we cover all possible returns?

**Review by zmaglica — CHANGES_REQUESTED** [review]
Good work on this PR. I left two comments about the edge cases.

File: includes/class-wc-payments-webhook-processing-service.php (hunk 1 above)
Comment (mgascam) [reply includes/class-wc-payments-webhook-processing-service.php:748]: Thanks for your suggestion to use the `WC_Abstract_Order` class @zmaglica. For context,  I was initially using [it](https://github.com/Automattic/woocommerce-payments/pull/9761/commits/b93afec06e5e94c8a8d53da101f6c0afdba02b30), but decided to go with the current approach after thinking that this flow is based on the `succeeded` and `failed` events of the `payment_intent`. In theory, we should never find an instance of `WC_Order_Refund`, but better to be safe than sorry. I'm going with the stricter approach 👍

Comment (mgascam) [reply includes/class-wc-payments-webhook-processing-service.php:748]: After thinking about it more, I believe it’s better not to use `WC_Abstract_Order` here. I get that we could potentially be dealing with a `WC_Order_Refund`, since that’s a possible return type of `this->wcpay_db->order_from_order_id`. But I don’t think that’s something we should expect, and I feel like throwing an exception would make more sense in this case.

Here’s my reasoning: we’re handling a payment_intent event from Stripe. If we don’t find an order linked to the `payment_intent` ID but end up with a `WC_Order_Refund` instead, I think it would be wrong to move forward with processing the refund object. Raising an exception feels like the right approach here.

Also, refunds in Stripe are their own separate entities and are dealt with in a different flow, so this situation doesn’t align with what we’d expect. That’s my take on it, but I’m happy to hear other thoughts or be corrected if I’m missing something.

File: tests/unit/test-class-wc-payments-webhook-processing-service.php (hunk 2 above)
Comment (mgascam) [reply tests/unit/test-class-wc-payments-webhook-processing-service.php]: Added a test here [cfb7801](https://github.com/Automattic/woocommerce-payments/pull/9761/commits/cfb7801f4e5068f3d2559db00d37ce3b1bc7dd27) but please check my previous comment.

File: includes/class-wc-payments-webhook-processing-service.php (hunk 1 above)
Comment (zmaglica) [reply includes/class-wc-payments-webhook-processing-service.php:748]: Got it. Let's stick with your original plan. If another type of WC_Order is encountered, an exception will be thrown, which is ideal since no other order types should exist.

**Review by zmaglica — APPROVED** [review]
LGTM! :shipit:
